### PR TITLE
[CALCITE-2580] RexSimplify: coalesce(null > null, true) produces wron…

### PR DESCRIPTION
**Pull request by pengzhiwei2018**
…g result in unknownAsFalse mode

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/rex/RexSimplify.java [hunk 1]
```diff
@@ -578,8 +578,10 @@ private RexNode simplifyCoalesce(RexCall call) {
     final Set<String> digests = new HashSet<>();
     final List<RexNode> operands = new ArrayList<>();
     for (RexNode operand : call.getOperands()) {
-      operand = simplify_(operand);
-      if (digests.add(operand.digest)) {
+      operand = withUnknownAsFalse(false).simplify_(operand);
+      if (digests.add(operand.digest)
+          && (!RexUtil.isNull(operand)
+                  || !unknownAsFalse)) {
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/rex/RexSimplify.java]: RexUtil.isNull and unknownAsFalse look very odd here

Comment (pengzhiwei2018) [reply core/src/main/java/org/apache/calcite/rex/RexSimplify.java]: well,In unknownAsFalse mode,Coalesce(null,x) will simplify to x

File: core/src/main/java/org/apache/calcite/rex/RexSimplify.java [hunk 2]
```diff
@@ -577,9 +577,12 @@ private RexNode simplifyIsNull(RexNode a) {
   private RexNode simplifyCoalesce(RexCall call) {
     final Set<String> digests = new HashSet<>();
     final List<RexNode> operands = new ArrayList<>();
+    final RexSimplify simplify = withUnknownAsFalse(false);
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/rex/RexSimplify.java]: Is this really needed for CALCITE-2580?

Comment (pengzhiwei2018) [reply core/src/main/java/org/apache/calcite/rex/RexSimplify.java]: In UnknownAsFalse mode,for "Coalesce(null,x)",  "null" will simplify to "false" without this code which lead to the wrong result.

Comment (vlsi) [reply core/src/main/java/org/apache/calcite/rex/RexSimplify.java]: makes sense

File: core/src/main/java/org/apache/calcite/rex/RexSimplify.java [hunk 3]
```diff
@@ -577,9 +577,12 @@ private RexNode simplifyIsNull(RexNode a) {
...
     for (RexNode operand : call.getOperands()) {
-      operand = simplify_(operand);
-      if (digests.add(operand.digest)) {
+      operand = simplify.simplify_(operand);
+      if (digests.add(operand.digest)
+          && (!RexUtil.isNull(operand)
+                  || !unknownAsFalse)) {
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/rex/RexSimplify.java]: Are those `!RexUtil.isNull(operand)|| !unknownAsFalse` checks really required?

Comment (pengzhiwei2018) [reply core/src/main/java/org/apache/calcite/rex/RexSimplify.java]: In unknownAsFalse mode,without this check,the "Coalesce(null,x)" will not simplify to "x".I can  
 change this to "!(RexUtil.isNull(operand)&& unknownAsFalse)" which is more understandable.

Comment (vlsi) [reply core/src/main/java/org/apache/calcite/rex/RexSimplify.java]: This is an optimization feature that is not listed in CALCITE-2580, neither it is covered in the tests.

I suggest to trim "leading nulls" to a separate JIRA.

Comment (pengzhiwei2018) [reply core/src/main/java/org/apache/calcite/rex/RexSimplify.java]: Ok

**Pull request closed without merging**